File: leetcode/1456.maximum-number-of-vowels-in-a-substring-of-given-length.cpp
/*
 * @lc app=leetcode id=1456 lang=cpp
 *
 * [1456] Maximum Number of Vowels in a Substring of Given Length
 */

// @lc code=start

#define IS_VOWEL(x) (x == 'a' || x == 'e' || x == 'i' || x == 'o' || x == 'u')

class Solution {
public:
	int maxVowels(string s, int k) {
		int	len = s.size();
		int	vowel = 0;
		int maxVowel = 0;

		for (int i = 0; i < len; i++) {
			if (i < k) {
				if (IS_VOWEL(s[i])) {
					vowel++;
					maxVowel++;
				}
				continue ;
			}
			vowel += IS_VOWEL(s[i]) - IS_VOWEL(s[i - k]);
			maxVowel = max(vowel, maxVowel);
			if (maxVowel == k) // faster for specific cases
				return k ;
		}
		return maxVowel;
	}
};
// @lc code=end

File: leetcode/1456.maximum-number-of-vowels-in-a-substring-of-given-length.py
#
# @lc app=leetcode id=1456 lang=python3
#
# [1456] Maximum Number of Vowels in a Substring of Given Length
#

# @lc code=start
class Solution:
	def maxVowels(self, s: str, k: int) -> int:
		vowels, maxVowels = 0, 0
		for i in range(len(s)):
			if (i < k):
				if s[i] in {'a', 'e', 'i', 'o', 'u'}:
					vowels += 1
					maxVowels += 1
			else:
				vowels += (s[i] in {'a', 'e', 'i', 'o', 'u'}) - (s[i - k] in {'a', 'e', 'i', 'o', 'u'})
				maxVowels = max(vowels, maxVowels)
		return maxVowels
# @lc code=end
